apm-agent-tracer/src/main/java/co/elastic/apm/agent/tracer/ElasticContext.java

```diff
@@ -18,26 +18,113 @@
  */
 package co.elastic.apm.agent.tracer;
 
-public interface ElasticContext<T extends ElasticContext<T>> {
+import co.elastic.apm.agent.tracer.dispatch.BinaryHeaderSetter;
+import co.elastic.apm.agent.tracer.dispatch.TextHeaderGetter;
+import co.elastic.apm.agent.tracer.dispatch.TextHeaderSetter;
+import co.elastic.apm.agent.tracer.reference.ReferenceCounted;
+
+import javax.annotation.Nullable;
+
+public interface ElasticContext<T extends ElasticContext<T>> extends ReferenceCounted, Activateable<T> {
```

Comment:
`ElasticContext` is now reference counted. The reason is that context can reference spans and spans are reference counted.
E.g. if you were to propagate a context to another thread, you need to make sure to do an `incrementReferenceCount` on the context, just like we did for `Spans` until now.